test/framework/control_plane.go

```diff
@@ -69,81 +77,109 @@ func (input *ControlplaneClusterInput) ControlPlaneCluster() {
 		}
-		return err
-	}, input.CreateTimeout, 10*time.Second).Should(BeNil())
+		return nil
+	}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
+
+	By("creating related resources")
+	for _, obj := range input.RelatedResources {
+		By(fmt.Sprintf("creating a/an %s resource", obj.GetObjectKind().GroupVersionKind()))
+		Eventually(func() error {
+			return mgmtClient.Create(ctx, obj)
+		}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
+	}
+
+	// expectedNumberOfNodes is the number of nodes that should be deployed to
+	// the cluster. This is the control plane nodes plus the number of replicas
+	// defined for a possible MachineDeployment.
+	expectedNumberOfNodes := len(input.Nodes)
+
+	// Create the additional control plane nodes.
+	for i, node := range input.Nodes {
+		expectedNumberOfNodes++
 
-	// create all the machines at once
-	for _, node := range input.Nodes {
-		By("creating an InfrastructureMachine resource")
+		By(fmt.Sprintf("creating %d control plane node's InfrastructureMachine resource", i+1))
 		Expect(mgmtClient.Create(ctx, node.InfraMachine)).To(Succeed())
 
-		By("creating a bootstrap config")
+		By(fmt.Sprintf("creating %d control plane node's BootstrapConfig resource", i+1))
 		Expect(mgmtClient.Create(ctx, node.BootstrapConfig)).To(Succeed())
 
-		By("creating a core Machine resource with a linked InfrastructureMachine and BootstrapConfig")
+		By(fmt.Sprintf("creating %d control plane node's Machine resource with a linked InfrastructureMachine and BootstrapConfig", i+1))
 		Expect(mgmtClient.Create(ctx, node.Machine)).To(Succeed())
 	}
 
-	// Wait for the cluster infrastructure
-	Eventually(func() string {
+	By("waiting for cluster to enter the provisioned phase")
+	Eventually(func() (string, error) {
 		cluster := &clusterv1.Cluster{}
 		key := client.ObjectKey{
 			Namespace: input.Cluster.GetNamespace(),
 			Name:      input.Cluster.GetName(),
 		}
 		if err := mgmtClient.Get(ctx, key, cluster); err != nil {
-			return err.Error()
+			return "", err
 		}
-		return cluster.Status.Phase
-	}, input.CreateTimeout, 10*time.Second).Should(Equal(string(clusterv1.ClusterPhaseProvisioned)))
+		return cluster.Status.Phase, nil
+	}, input.CreateTimeout, eventuallyInterval).Should(Equal(string(clusterv1.ClusterPhaseProvisioned)))
 
-	// wait for all the machines to be running
-	By("waiting for all machines to be running")
-	for _, node := range input.Nodes {
-		Eventually(func() string {
-			machine := &clusterv1.Machine{}
-			key := client.ObjectKey{
-				Namespace: node.Machine.GetNamespace(),
-				Name:      node.Machine.GetName(),
-			}
-			if err := mgmtClient.Get(ctx, key, machine); err != nil {
-				return err.Error()
-			}
-			return machine.Status.Phase
-		}, input.CreateTimeout, 10*time.Second).Should(Equal(string(clusterv1.MachinePhaseRunning)))
+	// Create the machine deployment if the replica count >0.
+	if machineDeployment := input.MachineDeployment.MachineDeployment; machineDeployment != nil {
```

Comment:
I'm ok with this for now, but I want us to watch out for making this test do too much. This if statement has the ability to quickly grow into other Ifs and muddy this test. If there's a set up step we can establish and reuse that between different tests (control plane, machine deployment, etc), that would be great, but that's longterm planning and I'm totally cool with this going through but acknowledge we should watch out for adding too many ifs in this test.

Reply:
Hmmm, I kind of see your point? But I think what you're saying makes more sense before when the inputs were separate from the function that used them. In that design the inputs could be viewed as immutable, or at least immutable once copied by value into the functions that used them. 

In the refactor from https://github.com/kubernetes-sigs/cluster-api/commit/3efd98f2be0d39036971d602b2be2ebe7753aa4c I think it makes sense to support optional resources, of which a `MachineDeployment` is one such thing. Because now the input is kind of the progenitor of the resource that performs the actual test.

In other words, I liked that the functions for `ControlPlane` or `MultiNodeControlPlane`, etc. were functional in design. You could then design them to accept multiple input types that could be various CAPI deployment strategies. However, having one input that can do *everything*, even if optionally, seems to be in line with the current model.